diff --git a/surf.c b/surf.c
index af0fa74..bde897b 100644
--- a/surf.c
+++ b/surf.c
@@ -646,11 +646,11 @@ updatetitle(Client *c)
 		getpagestats(c);
 
 		if (c->progress != 100)
-			title = g_strdup_printf("[%i%%] %s:%s | %s",
-			        c->progress, togglestats, pagestats, name);
+			title = g_strdup_printf("[%i%%] %s",
+			        c->progress, name);
 		else
-			title = g_strdup_printf("%s:%s | %s",
-			        togglestats, pagestats, name);
+			title = g_strdup_printf("%s",
+			         name);
 
 		gtk_window_set_title(GTK_WINDOW(c->win), title);
 		g_free(title);
