From 9858964ec2b1eeffe9e112cc8a6272efdc3ea8c3 Mon Sep 17 00:00:00 2001
From: Alexis Ben Miloud--Josselin <alexis.bmj@protonmail.com>
Date: Thu, 24 May 2018 10:52:00 +0200
Subject: [PATCH] Do not set default chan when joinning

---
 sic.c | 2 --
 1 file changed, 2 deletions(-)

diff --git a/sic.c b/sic.c
index ecefaf2..b7efd7e 100644
--- a/sic.c
+++ b/sic.c
@@ -77,8 +77,6 @@ parsein(char *s) {
 		switch(c) {
 		case 'j':
 			sout("JOIN %s", p);
-			if(channel[0] == '\0')
-				strlcpy(channel, p, sizeof channel);
 			return;
 		case 'l':
 			s = eat(p, isspace, 1);
-- 
2.11.0

